include/exadg/operators/generalized_laplace_operator.h

```diff
@@ -0,0 +1,849 @@
+      unsigned int const                                q,
+      FaceIntegrator<dim, n_components, Number> const & integrator,
+      OperatorType const &                              operator_type,
+      BoundaryType const &                              boundary_type,
+      dealii::types::boundary_id const                  boundary_id,
+      std::shared_ptr<BoundaryDescriptor<dim> const>    boundary_descriptor,
+      double const                                      time)
+  {
+    if(boundary_type == BoundaryType::Dirichlet)
+      return coeff_times_gradient_times_normal_m;
+
+    if(boundary_type == BoundaryType::Neumann)
+    {
+      if(operator_type == OperatorType::full || operator_type == OperatorType::inhomogeneous)
+      {
+        auto const bc       = boundary_descriptor->neumann_bc.find(boundary_id)->second;
+        auto const q_points = integrator.quadrature_point(q);
+
+        auto const h = FunctionEvaluator<value_rank, dim, Number>::value(bc, q_points, time);
+
+        return coeff_times_gradient_times_normal_m + value_type(2.0 * h);
+      }
+      else if(operator_type == OperatorType::homogeneous)
+        return -coeff_times_gradient_times_normal_m;
+      else
+        AssertThrow(false, dealii::ExcNotImplemented());
+    }
+
+    AssertThrow(false, dealii::ExcMessage("Boundary type of face is invalid or not implemented."));
+
+    return value_type{};
+  }
+};
+} // namespace Boundary
+
+/**
+ * Collection of data for the generalized Laplace operator
+ */
+template<int dim>
+struct OperatorData : public OperatorBaseData
+{
+  KernelData<dim> kernel_data{};
+
+  std::shared_ptr<Boundary::BoundaryDescriptor<dim>> bc{};
+};
+
+/**
+ * Generalized Laplace operator
+ *
+ * This class provides the loop structures for the integrals that are evaluated for a generalized
+ * Laplace term.
+ *
+ * @tparam n_components Number of components of the solution (vector) field. Solution is scalar if
+ * this is equal to 1.
+ * @tparam coupling_coefficient Boolean switch to differentiate if the coefficient type couples
+ * the solution vector components or not.
+ *  - `false` always results in a scalar coefficient,
+ *  - `true` results in a 2-rank tensor coefficient if `n_components = 1` and in a 4-rank tensor
+ *  coefficient if `n_components > 1`.
+ *
+ */
+template<int dim, typename Number, int n_components = 1, bool coupling_coefficient = false>
+class Operator : public OperatorBase<dim, Number, n_components>
+{
+private:
+  using scalar = dealii::VectorizedArray<Number>;
+  using vector = dealii::Tensor<1, dim, scalar>;
+
+  static constexpr unsigned int value_rank = (n_components > 1) ? 1 : 0;
+  static constexpr unsigned int coefficient_rank =
+    (coupling_coefficient) ? ((n_components > 1) ? 4 : 2) : 0;
+
+  using value_type    = dealii::Tensor<value_rank, dim, scalar>;
+  using gradient_type = dealii::Tensor<value_rank + 1, dim, scalar>;
+
+  using coefficient_type = dealii::Tensor<coefficient_rank, dim, scalar>;
```

Comment:
can one ask the kernel for these types (to avoid this duplication of code)?

Reply:
I guess I could define them in the namespace as alias templates, so outside the classes instead of inside them 👍

Reply:
see comment https://github.com/exadg/exadg/pull/414#discussion_r1219851578